web-api/src/personalization/stateless.rs

```diff
@@ -0,0 +1,369 @@
+// Copyright 2023 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use std::collections::HashMap;
+
+use chrono::{DateTime, Utc};
+use itertools::Itertools;
+use serde::Deserialize;
+use xayn_ai_bert::NormalizedEmbedding;
+use xayn_ai_coi::{CoiSystem, UserInterests};
+
+use super::PersonalizationConfig;
+use crate::{
+    error::{common::HistoryTooSmall, warning::Warning},
+    models::{DocumentId, DocumentTag},
+    storage::{self, TagWeights},
+    Error,
+};
+
+#[derive(Deserialize)]
+pub(super) struct UnvalidatedHistoryEntry {
```

Comment:
💅 Just a suggestion, feel free to dismiss, unvalid is not a correct word (but we get the meaning of course).
Consider `CandidateHistoryEntry` maybe

Reply:
The thing is it's not really a candidate it's a fixed entry we will take, it's just not yet clear if the user gave us malformed data.

Tbh. I'm not support happy about how things are handled wrt. validation  in general (not specific to my code).

The general problem is that when type safety and contextual enforcement of soft validity constraints overlap it can easily lead to very verbose code, and it did in our case.

In more detail: For many of the types like e.g. `DocumentId` we "soft" enforce the id's validity. What I mean with it is that if we get the id from a database (PG,ES) we take _whatever we get without validation_. Because there is just no sane way to handle the situation where a id in out system is "malformed", we can't just stop operating our service nor can we just randomly delete it or anything. (Side note: Because of stuff like that it's important to always handle encoding of queriers appropriately even if that shouldn't be needed given our constraints for what is a valid id.)

But when getting a id as input we want to always make sure it is valid.

The problem here is there is no (non magic) way to implicitly  pass the "context" under which the `DocumentId` is created when creating it as part of other structs through a serde `Deserializer` (or `FromRow` in PG, etc.).

In case of PG we kinda can just argue it's only used for internal databases (in our use-case) so the context is always the same so we can just implicitly assume it.

But for data from e.g. ES and inputs from JSON this doesn't work as they all go through serde/JSON.

So we had a few ways to change this:

- have some alternative serde impl for some of the interfaces using a wrapper and "external" annotations and mirror types

- parse a different type and validate by hand (easier the the first point, it's what we did)

- use some magic to pass in a context in some way

Additionally the validation is dependent on the config, too in some cases.

And while we went with the second method I'm not sure it was the best idea,  through also there is no off the shelf solution I'm aware of for the 3rd solution and creating and maintaining one isn't the best investment of time for the current situation :sweat_smile: